### DISCUSSION TITLE
关于WPF的VisualTree、事件触发源与事件路由

### DISCUSSION THREAD

**Discussion by AlvinRey** [BYJRK/DotNet-Discussions#187] [WPF 开发]
- 案例1：现在有一个需求，在一个ListBoxItem内有一个删除按钮，当我点击这个按钮时，删除按钮所在的Item。

所以关键是点击按钮时，ListBox要知道所触发的按钮是在哪个Item内。
我目前的解决方法是在ListBox上通过AddHandler，添加对鼠标点击事件的处理。在响应函数里，先通过OriginalSource得到VisualTree上触发源，再沿着VisualTree向上寻找，直到找到Item。最后把这个Item设置为Selected。
![image](https://github.com/BYJRK/DotNet-Discussions/assets/56423507/84c3fb31-87f8-4008-8baa-7cb4de669a09)
这里我的疑问是：只能通过这种循环的方式向上寻找吗？有没有更好的方法呢？还是说要自己维护一个字典去存储对应关系？

- 案例2：ListBox嵌套，就是ListBox的Item里面是另外一个ListBox。
先解释一下为什么要用ListBox嵌套而不用Tree。因为我能确定只有两层结构，而Tree是用来应对不确定的层数的，所以我觉得没必要用Tree。
我的疑问是关于SelectionChanged这个事件的，为什么当更改内层ListBox的所选项时，外层的SelectionChanged也会被触发？其中的原理是什么。

**Comment by BYJRK** [comment]
第一个问题，如果你不使用 MVVM 的话，你或许可以考虑在 Button 的实例化阶段就直接用闭包的方式为它注册 Click 事件，并传入合适的引用
第二个问题我没试过，不过听起来应该和 WPF 的冒泡事件有关？

**Reply by AlvinRey** [reply]
对于我目前对wpf的了解程度来说，Button的实例化是没法干预的。大佬能举一些例子，如何自定义Button的实例化吗？
我的认知里，写在Xaml里面的最终要先通过特殊的转换器转换成C#代码，与后置代码的类一起合并成一个最终的完整的类。但是转换的结果应该不是我能控制的吧。
如果说在后置代码中直接New一个Button，这样我还能理解“在 Button 的实例化阶段就直接用闭包的方式为它注册 Click 事件”这句话。。

**Reply by AlvinRey** [reply]
就算可以，但是我有使用MVVM。。为什么这个跟mvvm有关系，是mvvm涉及数据绑定？Item都是Template生成的？如果是这样的话，我这个案例就是了，ListBox的Item是根据一个Collection生成的

**Comment by toonbear** [comment]
搞复杂了吧,wpf是数据驱动的,直接删数据不就行了吗
我也是初学者,这个功能我刚好前天写了的,在listbox中显示一个最近打开文件的集合,每个item由文本和按钮组成,点击按钮删除一条记录
